text/0047-unspecified-semantics.md

```diff
@@ -0,0 +1,178 @@
+`User::"alice" == principal`, those expressions currently (Cedar 2.x and 3.x)
+evaluate to `false` -- unspecified entities are defined to not be equal to any
+entity in the hierarchy.
+2. If some policy contains the expression `principal in Group::"foo"` or
+`User::"alice" in principal`, those expressions currently (Cedar 2.x and 3.x)
+evaluate to `false` -- unspecified entities are defined to not be `in` any
+entity in the hierarchy.
+3. If some policy contains the expression `principal == principal` or
+`principal in principal`, those expressions currently (Cedar 2.x and 3.x)
+evaluate to `true` (even when `principal` is unspecified in the request).
+Unspecified entities are defined to be equal to themselves.
+4. If some policy contains the expression `principal has foo`, that expression
+currently (Cedar 2.x and 3.x) evaluates to `false` -- unspecified entities are
+defined to have no attributes.
+4. If some policy contains the expression `principal.foo`, that expression
+currently throws an error.
+
+With this RFC, we change the behavior to be consistent: _All_ of the above
+examples will throw an error.
+
+## Motivation
+
+The behavior of unspecified entities is complicated, unintuitive, and has to
+be special-cased frequently in the validator.  Even we, as Cedar developers, are
+often confused or tripped up by the semantics of `Unspecified`.
+The distinction between unspecified entities and partial-evaluation unknowns is
+also subtle and confusing and has caused bugs.
+
+The proposed semantics is more consistent for Cedar users, as outlined above in
+"Basic example", and hopefully also more intuitive for users (reduce surprises).
+The original intended use-case for `Unspecified` was that a caller who leaves a
+variable (say `principal`) unspecified, is implicitly guaranteeing that
+`principal` does not matter to the authorization outcome (e.g., because no
+policies check `principal`).
+This RFC's proposed semantics is more true to this intended use-case than the
+current semantics.
+
+Additionally, the proposed semantics will allow us to simplify Cedar's
+implementation by unifying the behavior of `Unspecified` and `Unknown` in many
+ways, allowing us to share code for these two features.
+
+## Detailed design
+
+### Semantics
+
+The detailed semantics is clear and simple: if `principal` (resp. `action` or
+`resource`) is Unspecified in the request, it is an evaluation error for
+`principal` (resp. `action` or `resource`) to appear in any (evaluated)
+expression in the policy.
+(The scope constraints `principal,` `action,` and `resource,` are always still
+allowed of course.)
+
+From the evalautor perspective, we must still allow unspecified variables to
+appear in unevaluated/dead code, because the evaluator simply does not evaluate
+those expressions due to short-circuiting; but despite this, validator can still
+soundly prohibit these variables in all expressions in the policy.
```

Comment:
I don't understand what this means. I think you mean that if we evaluate `if true then true else principal has owner` and `principal` is Unspecified, this will still evaluate to `true`. I feel like this goes without saying?

Reply:
You are correct that's what it means.  Perhaps it goes without saying, but because of the language "appears in the policy" in the previous policy, I thought it was safer to clarify.  This is a Detailed Design section after all

Reply:
It's a little less clear whether your snippet should still validate.  My text points out that we could write the validator to produce an error for your snippet.  We could also write the validator to validate your snippet.  Both are sound